wpilibj/src/main/java/edu/wpi/first/wpilibj/RobotBase.java

```diff
@@ -221,44 +219,28 @@ public static void initializeHardwareConfiguration() {
    */
   @SuppressWarnings({"PMD.AvoidInstantiatingObjectsInLoops", "PMD.AvoidCatchingThrowable",
                      "PMD.CyclomaticComplexity", "PMD.NPathComplexity"})
-  public static void main(String... args) {
+  public static <T extends RobotBase> void startRobot(Supplier<T> robotSupplier) {
     if (!HAL.initialize(500, 0)) {
       throw new IllegalStateException("Failed to initialize. Terminating");
     }
 
     HAL.report(tResourceType.kResourceType_Language, tInstances.kLanguage_Java);
 
-    String robotName = "";
-    if (args.length > 0) {
-      robotName = args[0];
-    } else {
-      Enumeration<URL> resources = null;
-      try {
-        resources = Thread.currentThread()
-            .getContextClassLoader().getResources("META-INF/MANIFEST.MF");
-      } catch (IOException ex) {
-        ex.printStackTrace();
-      }
-      while (resources != null && resources.hasMoreElements()) {
-        try {
-          Manifest manifest = new Manifest(resources.nextElement().openStream());
-          robotName = manifest.getMainAttributes().getValue("Robot-Class");
-        } catch (IOException ex) {
-          ex.printStackTrace();
-        }
-      }
-    }
-
     System.out.println("********** Robot program starting **********");
 
-    RobotBase robot;
+    T robot;
     try {
-      robot = (RobotBase) Class.forName(robotName).newInstance();
+      robot = robotSupplier.get();
     } catch (Throwable throwable) {
       Throwable cause = throwable.getCause();
       if (cause != null) {
         throwable = cause;
```

Comment:
Is this necessary? Any exception thrown will occur in user code.

Reply:
Yeah but we want to catch it and print it properly to the DS before we let the crash finish.